### PR TITLE
Add the identity function as core::convert::identity

### PR DESCRIPTION
**Pull request by Centril**
## New notes

This implements rust-lang/rfcs#2306 (see https://github.com/rust-lang/rust/issues/53500).

## Old notes (ignore this in new reviews)

Adds the identity function `fn id<T>(x: T) -> T { x }` to core::convert and the prelude.
Some motivations for why this is useful are explained in the doc tests.
Another is that using the identity function instead of `{ x }` or `|x| x` makes it clear that you intended to use an identity conversion on purpose.

The reasoning:
+ behind adding this to `convert` and not `mem` is that this is an identity *conversion*.
+ for adding this to the prelude is that it should be easy enough to use that the ease of writing your own identity function or using a closure `|x| x` doesn't overtake that.

I've separated this out into two feature gates so that the addition to the prelude can be considered and stabilized separately.

cc @bluss

### CODE REVIEW

**Comment by Manishearth** [comment]
I'm wary of adding unstable stuff to the prelude, feels like it could still cause clashes with other glob imports.

(also, shouldn't this go through an rfc first?)

**Comment by Centril** [comment]
> I'm wary of adding unstable stuff to the prelude, feels like it could still cause clashes with other glob imports.

Any example you have in mind? AFAIK any import or function definition that the user makes, glob or not will override anything in the prelude..

> (also, shouldn't this go through an rfc first?)

I can do that if you want...

But the addition is rather minor - there's not much technical to discuss - only if we want this in libcore or not... And there's precedent for this in other languages like [Haskell](http://hackage.haskell.org/package/base-4.10.1.0/docs/Prelude.html#v:id), [Scala](https://www.scala-lang.org/api/current/scala/Predef$.html#identity[A](x:A):A), [Java](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html#identity--), [Idris](https://www.idris-lang.org/docs/1.0/prelude_doc/docs/Prelude.Basics.html)

**Comment by Manishearth** [comment]
@Centril the interaction between two glob imports is the thing here.

Adding stuff to the stdlib usually goes through an RFC, however trivial. This goes double for things in the prelude. It's sometimes fine to add stuff as a feature-gated API before making the RFC but you need some justification for that.

**Comment by Centril** [comment]
@Manishearth Aight; I'll write up an RFC. I'll leave this PR open meanwhile so that we can use that one if the RFC gets accepted.

**Comment by Manishearth** [comment]
Probably should close this and reopen if the RFC merges. Unless we intend to get it merged soon, it shouldn't stay open.

**Comment by Centril** [comment]
That's fair. Expect and RFC filed by the end of the day.

**Comment by bluss** [comment]
The prelude change seems good to have an rfc for? New minor features do not otherwise need an rfc, I think this change fits well inside what has been added as unstable without rfc recently.

**Comment by Centril** [comment]
@bluss Yeah. Tho I think that in this case the function is much less useful outside the prelude.

**Comment by Centril** [comment]
Once the tests pass, this should be ready to go.

I've added tests for the `rustc_const_unstable(..)` bit as well.

File: src/libcore/convert.rs [hunk 1]
```diff
@@ -48,6 +48,66 @@
...
+///
+/// ```rust
+/// #![feature(convert_id)]
+/// use std::convert::identity;
+///
+/// let iter = vec![Some(1), None, Some(3)].into_iter();
+/// let filtered = iter.filter_map(identity).collect::<Vec<_>>();
+/// assert_eq!(vec![1, 3], filtered);
+/// ```
+#[unstable(feature = "convert_id", issue = "0")]
```
Comment (varkor) [inline src/libcore/convert.rs]: Should be `issue = "53500"` now.

Comment (Centril) [reply src/libcore/convert.rs]: Cheers; you always miss something :D

**Pull request merged**